kedro/framework/cli/catalog.py

```diff
@@ -207,3 +207,50 @@ def rank_catalog_factories(metadata: ProjectMetadata, env):
         click.echo(yaml.dump(list(catalog_factories.keys())))
     else:
         click.echo("There are no dataset factories in the catalog.")
+
+
+@catalog.command("resolve")
+@env_option
+@click.pass_obj
+def resolve_patterns(metadata: ProjectMetadata, env):
+    """Resolve pipeline datasets with catalog factories."""
+
+    session = _create_session(metadata.package_name, env=env)
+    context = session.load_context()
+
+    data_catalog = context.catalog
+    catalog_config = context.config_loader["catalog"]
+
+    catalog_config = {
+        ds_name: ds_config
+        for ds_name, ds_config in catalog_config.items()
+        if not data_catalog._is_pattern(ds_name)
+    }
+
+    target_pipelines = pipelines.keys()
+
+    for pipe in target_pipelines:
```

Comment:
Could these nested loops be separated into two? 
* For all pipelines in `target_pipelines` add their datasets to a set
* Then for each `ds_name` in that set do the matching

Reply:
Definitely possible, but I am not able to see an advantage to the change? From my understanding the nested loops give us a time complexity of **number-of-pipelines * datasets-per-pipeline**, or **n * m/n** (if we assume an average distribution of datasets per pipeline, in any case we would still only iterate through m datasets regardless of distribution). This can be simplified to a linear time complexity just **m** of w.r.t the number of datasets. Having two separate loops gives us a time of **n+m** instead.

Is the aim just to move the matching logic outside of the nested loop? It would still be run on each dataset so I don't really see any optimisation to be gained from the change 🤔

Reply:
The reasoning for this being it would make the code easier to read. The factory datasets are only matched and added once in the current implementation, and they will be skipped in consequent checks so the time complexity wouldn't be affected much. This still loops through all datasets for all pipelines - which means all datasets are being checked at least twice (once for their own pipeline + once for the default pipeline + more if they belong to multiple pipelines)

Reply:
That makes sense! Changed in [8784a37](https://github.com/kedro-org/kedro/pull/2891/commits/8784a377b35135c7b36933f89737b682ab4c879e)